### ISSUE TITLE
Incorrect name resolution when __init__ contains variable with same name as a module

### ISSUE THREAD

**Issue by xsrvmy** [microsoft/pyright#3532]
**Describe the bug**
When __init__ contains a variable with the same name as a module in the package, python resolves the import to the module, but pyright resolves it to the variable and reports type errors

**To Reproduce**
See code

**Expected behavior**
The behaviour matches python.

**Screenshots or Code**
a/\_\_init\_\_.py
```python
b = []
```
a/b.py
```python
d = 3
```
main.py
```python
import a.b
from a import b

print(b.d) # python prints 3, pyright gives Cannot access member "d" for type "list[Unknown]"
print(a.b.d) # python prints 3, pyright does not error
```

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**VS Code extension or command-line**
Able to reproduce with Pyright 1.1.239 in codium and coc-pyright 1.1.252 in neovim. Also same problem occurs with PyLance 2022.6.0 (in the code above, b. will hint as if b is a list).

**Additional context**
Workaround: `import a.b as b` doesn't cause this issue

**Comment by erictraut** [comment]
What you're doing here is very fragile and is not possible to model statically in the general case. You are relying on temporal-ordering side effects of the module loading mechanism in Python.

To demonstrate, change `__init__.py` in your example to the following:
```python
# __init__.py
from .b import d
b = []
```

Now run `main.py` and note that it crashes because `b` is a list.

Next, reverse the order of the two lines in `__init__.py`:
```python
# __init__.py
b = []
from .b import d
```
Now your example works without crashing.

The reason this works is that `b` is being overwritten in the namespace of the `__init__.py` module at different times depending on when the submodule `b` happens to be imported. 

To demonstrate how fragile this is, let's introduce another submodule to the mix.
```python
# __init__.py
from .c import d
b = []
```
```python
# c.py
from .b import d
```

Now run `main.py`, and it crashes. That's because the `c` causes `b` to be imported and cached by the loader and written to `__init__.py`'s namespace. Then `b` is overwritten by a list. When you later attempt to import `b` again by `main.py`, the cached version is returned rather than overwriting `a.b` with the module.

Pyright doesn't attempt to model these temporal side effects of the loader because, in practice, they cannot be predicted statically. They depend on the order in which code is executed at runtime.

You should avoid fragilities like this within your code. If you do, static type checking will work fine, and you won't run into unpleasant surprises at runtime.
